db/migrate/20231210061330_add_created_at_to_fts_targets_with_index.rb

```diff
@@ -0,0 +1,109 @@
+
+    ActiveRecord::Base.transaction do
+      update_created_at_using_last_modified_at
+      update_created_at_using_created_on
+    end
+
+    if Redmine::Database.mysql?
+      add_index :fts_targets, :created_at
+    else
+      remove_index :fts_targets, name: "fts_targets_index_pgroonga"
+      add_index :fts_targets,
+                [:id,
+                 :source_id,
+                 :source_type_id,
+                 :project_id,
+                 :container_id,
+                 :container_type_id,
+                 :custom_field_id,
+                 :is_private,
+                 :last_modified_at,
+                 :created_at,
+                 :title,
+                 :content,
+                 :tag_ids],
+                using: "PGroonga",
+                with: "normalizer = 'NormalizerNFKC121'",
+                name: "fts_targets_index_pgroonga"
+    end
+  end
+
+  def down
+    return if !table_exists?(:fts_targets)
+
+    if Redmine::Database.mysql?
+      remove_index :fts_targets, :created_at
+    else
+      remove_index :fts_targets, name: "fts_targets_index_pgroonga"
+      add_index :fts_targets,
+                [:id,
+                 :source_id,
+                 :source_type_id,
+                 :project_id,
+                 :container_id,
+                 :container_type_id,
+                 :custom_field_id,
+                 :is_private,
+                 :last_modified_at,
+                 :title,
+                 :content,
+                 :tag_ids],
+                using: "PGroonga",
+                with: "normalizer = 'NormalizerNFKC121'",
+                name: "fts_targets_index_pgroonga"
+    end
+
+    remove_column :fts_targets, :created_at
+  end
+
+  private
+
+  class FtsTarget < ActiveRecord::Base
+  end
+
+  class FtsType < ActiveRecord::Base
+  end
+
+  class Issue < ActiveRecord::Base
+  end
+
+  class Message < ActiveRecord::Base
+  end
+
+  class Project < ActiveRecord::Base
+  end
+
+  class WikiPage < ActiveRecord::Base
+  end
+
+  def update_created_at_using_last_modified_at
+    types = ['Attachment', 'Change', 'Changeset', 'CustomValue', 'Document', 'Journal', 'News']
+
+    FtsTarget.where(source_type_id: FtsType.where(name: types).select(:id))
+             .update_all(created_at: FtsTarget.arel_table[:last_modified_at])
+  end
+
+  def update_created_at_using_created_on
+    types = ['Issue', 'Message', 'Project', 'WikiPage']
+
+    types.each do |type_name|
+      source_table = type_name.constantize.arel_table
+      fts_targets_table = FtsTarget.arel_table
+
+      subquery_for_created_on = source_table.project(source_table[:created_on])
+                                            .where(source_table[:id].eq(fts_targets_table[:source_id]))
+                                            .to_sql
+
+      FtsTarget.where(source_type_id: FtsType.where(name: type_name).select(:id))
+               .update_all("created_at = (#{subquery_for_created_on})")
```

Comment:
Can we use `Arel.sql` here?

```suggestion
               .update_all(created_at: Arel.sql(subquery_for_created_on))
```

Reply:
I attempted it, but the syntax was not interpreted as we anticipated.

<details>
<summary>Error logs</summary>
```console
> query = Issue.where(Issue.arel_table[:id].eq(FtsTarget.arel_table[:source_id])).select(:created_on).to_sql
> FtsTarget.where(source_type_id: FtsType.where(name: Issue.name).select(:id)).update_all(created_at: Arel.sql(query))
/Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-6.1.7.6/lib/active_record/connection_adapters/postgresql_adapter.rb:672:in `exec_params': PG::SyntaxError: ERROR:  syntax error at or near "SELECT" (ActiveRecord::StatementInvalid)
LINE 1: UPDATE "fts_targets" SET "created_at" = SELECT "issues"."cre...
                                                ^
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-6.1.7.6/lib/active_record/connection_adapters/postgresql_adapter.rb:672:in `block (2 levels) in exec_no_cache'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.6/lib/active_support/dependencies/interlock.rb:48:in `block in permit_concurrent_loads'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.6/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.6/lib/active_support/dependencies/interlock.rb:47:in `permit_concurrent_loads'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-6.1.7.6/lib/active_record/connection_adapters/postgresql_adapter.rb:671:in `block in exec_no_cache'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-6.1.7.6/lib/active_record/connection_adapters/abstract_adapter.rb:696:in `block (2 levels) in log'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.6/lib/active_support/concurrency/load_interlock_aware_monitor.rb:26:in `block (2 levels) in synchronize'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.6/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `handle_interrupt'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.6/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `block in synchronize'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.6/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `handle_interrupt'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.6/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `synchronize'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-6.1.7.6/lib/active_record/connection_adapters/abstract_adapter.rb:695:in `block in log'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.6/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-6.1.7.6/lib/active_record/connection_adapters/abstract_adapter.rb:687:in `log'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-6.1.7.6/lib/active_record/connection_adapters/postgresql_adapter.rb:670:in `exec_no_cache'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-6.1.7.6/lib/active_record/connection_adapters/postgresql_adapter.rb:649:in `execute_and_clear'
        ... 69 levels...
/Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-6.1.7.6/lib/active_record/connection_adapters/postgresql_adapter.rb:672:in `exec_params': ERROR:  syntax error at or near "SELECT" (PG::SyntaxError)
LINE 1: UPDATE "fts_targets" SET "created_at" = SELECT "issues"."cre...
                                                ^
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-6.1.7.6/lib/active_record/connection_adapters/postgresql_adapter.rb:672:in `block (2 levels) in exec_no_cache'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.6/lib/active_support/dependencies/interlock.rb:48:in `block in permit_concurrent_loads'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.6/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.6/lib/active_support/dependencies/interlock.rb:47:in `permit_concurrent_loads'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-6.1.7.6/lib/active_record/connection_adapters/postgresql_adapter.rb:671:in `block in exec_no_cache'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-6.1.7.6/lib/active_record/connection_adapters/abstract_adapter.rb:696:in `block (2 levels) in log'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.6/lib/active_support/concurrency/load_interlock_aware_monitor.rb:26:in `block (2 levels) in synchronize'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.6/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `handle_interrupt'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.6/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `block in synchronize'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.6/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `handle_interrupt'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.6/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `synchronize'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-6.1.7.6/lib/active_record/connection_adapters/abstract_adapter.rb:695:in `block in log'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.6/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-6.1.7.6/lib/active_record/connection_adapters/abstract_adapter.rb:687:in `log'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-6.1.7.6/lib/active_record/connection_adapters/postgresql_adapter.rb:670:in `exec_no_cache'
        from /Users/otegami/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-6.1.7.6/lib/active_record/connection_adapters/postgresql_adapter.rb:649:in `execute_and_clear'
        ... 69 levels...
```

</details>

Reply:
`.update_all(created_at: Arel.sql("(#{subquery_for_created_on})"))`?

Reply:
fix: [a053743](https://github.com/clear-code/redmine_full_text_search/pull/118/commits/a053743d4a9fe3e5440160bc745bc076b3479145) Thank you so much. I missed it and your suggestion works.